data_safe_haven/pulumi/pulumi_stack.py

```diff
@@ -198,17 +198,15 @@ def login(self) -> None:
             except DataSafeHavenPulumiException:
                 AzureCli().login()  # this is needed to read the encryption key from the keyvault
                 env_vars = " ".join([f"{k}='{v}'" for k, v in self.env.items()])
-                command = f"pulumi login 'azblob://{self.cfg.pulumi.storage_container_name}'"
-                with subprocess.Popen(
-                    f"{env_vars} {command}",
-                    shell=True,
+                process = subprocess.run(
+                    ["pulumi", "login", f"'azblob://{self.cfg.pulumi.storage_container_name}'"],
+                    env=env_vars,
                     cwd=self.work_dir,
-                    stdout=subprocess.PIPE,
-                    stderr=subprocess.STDOUT,
+                    capture_output=True,
                     encoding="UTF-8",
-                ) as process:
-                    if process.stdout:
-                        self.logger.info(process.stdout.readline().strip())
+                )
+                process.check_returncode()
```

Comment:
Add a comment that this raises a `CalledProcessError` for a non-zero return code.

Reply:
Fixed in 4058ee0b